csharp/src/Microsoft.ML.OnnxRuntime/InferenceSession.shared.cs

```diff
@@ -1044,6 +1044,124 @@ public ulong ProfilingStartTimeNs
             }
         }
 
+        private static void CallbackWrapper(IntPtr userData, IntPtr[] outputs, uint numOutputs, IntPtr status)
+        {
+            var resourceHdl = GCHandle.FromIntPtr(userData);
+            CallbackResource resource = (CallbackResource)resourceHdl.Target;
+
+            try
+            {
+                // throw on error
+                NativeApiStatus.VerifySuccess(status);
+
+                if (numOutputs != resource.namedOutputValues.Count()) {
+                    throw new OnnxRuntimeException(ErrorCode.Fail, "number of output mismatch");
+                }
+
+                for (uint ith = 0; ith < numOutputs; ith++)
+                {
+                    if (null == outputs[ith]) { 
+                        throw new OnnxRuntimeException(ErrorCode.Fail, $"output '{ith}' is null");
+                    }
+                }
+
+                resource.userCallback(resource.userData, resource.namedOutputValues);
+            }
+            finally
+            {
+                resourceHdl.Free();
+            }
+        }
+
+        private delegate void CallbackWrapperDelegate(IntPtr userData, IntPtr[] outputs, uint numOutputs, IntPtr status);
+
+        private static CallbackWrapperDelegate callbackWrapper = new CallbackWrapperDelegate(CallbackWrapper);
+
+        /// <summary>
+        /// Function delegate for the callback
+        /// </summary>
+        /// <param name="userData">any arbitrary data that passed to RunAsync</param>
+        /// <param name="outputs">output tensors</param>
+        //public delegate void CallbackDelegate(IntPtr userData, IDisposableReadOnlyCollection<OrtValue> outputs);
+        public delegate void CallbackDelegate(IntPtr userData, IReadOnlyCollection<NamedOnnxValue> outputs);
```

Comment:
IIRC it would be more C#-ish to use `Action<T>` for a callback parameter instead of defining a delegate.

Reply:
code updated - no more delegate in public API.